airbyte-cdk/python/airbyte_cdk/sources/declarative/manifest_declarative_source.py

```diff
@@ -142,6 +144,33 @@ def _validate_source(self):
         except ValidationError as e:
             raise ValidationError("Validation against json schema defined in declarative_component_schema.yaml schema failed") from e
 
+        cdk_version = metadata.version("airbyte_cdk")
+        cdk_major, cdk_minor, cdk_patch = (int(part) for part in cdk_version.split("."))
+        manifest_version = self._source_config.get("version")
+        manifest_major, manifest_minor, manifest_patch = self._get_version_parts(manifest_version)
+
+        if cdk_major < manifest_major or (cdk_major == manifest_major and cdk_minor < manifest_minor):
+            raise ValidationError(
+                f"The manifest version {manifest_version} is greater than the airbyte-cdk package version ({cdk_version}). Your "
+                f"manifest may contain features that are not in the current CDK version."
+            )
+        elif manifest_major == 0 and manifest_minor < 29 and (0 < cdk_major or 29 <= cdk_minor):
+            raise ValidationError(
+                f"The low-code framework was promoted to Beta in airbyte-cdk version 0.29.0 and contains many breaking changes to the "
+                f"language. The manifest version {manifest_version} is incompatible with the airbyte-cdk package version "
+                f"{cdk_version} which contains these breaking changes."
+            )
+
+    @staticmethod
+    def _get_version_parts(version: str) -> (int, int, int):
```

Comment:
Should we document somewhere (maybe in `setup.py`) that if the versioning changes, this will be impacted? I remember some people talking about release candidate versions for example which would add things at the end of the version

Reply:
Very valid point, maybe using a lib like https://pypi.org/project/semver/ would be worth it - your call @brianjlai

Reply:
I do like the idea of not building a version comparator that might already exist, although my main concern is that in GH repo in particular they have at the top **"This is a development version. Do NOT use it in production before the final 3.0.0 is released."** 😟 

As far as I can tell they're still on the dev release so I'm not sure we want to quite rely on it just yet. I don't see too many alternatives or widely supported libraries. The other caveat that I can also see us simplifying our versioning notation on the manifest to just v1 or v1.2 and if we do that we might scrap using library anyway.

Interesting, I don't think I was part of those convos or maybe I just forgot. I think a comment in the CDK setup makes sense and I'll add that. And the other nice thing that is the feedback loop on this is very quick, any command run during testing will fail loudly with the error message so theres no way to check these connectors in w/o a fix and we should notice it during testing.